### PR TITLE
Cross-zone (rack) fallback implementation

### PR DESCRIPTION
**Pull request by jcacciatore**
Currently cross-zone fallback only occurs when dyno client recognizes that a dynomite server is down. With this change any time an error occurs either obtaining a connection from the pool or with the execution of an operation the connection for the retry will be borrowed from a remote zone rather than the same local host pool where the error just occurred.

This change also contains the following minor fixes:
* Bug fix in failover metric reporting (regression)
* Automatic disabling of local zone affinity if the local zone has not been set. This often happens w/local laptop development + unit testing.
* Cleanup of 'datacenter' terminology to rack/zone. Initially datacenter was equivalent to zone/rack in Dynomite terminology. This changed however the dyno client code was never changed to reflect this which makes reading the code confusing. Cleanup is not complete; there is more to do.

### CODE REVIEW

File: dyno-core/src/main/java/com/netflix/dyno/connectionpool/impl/ConnectionPoolConfigurationImpl.java [hunk 1]
```diff
@@ -62,9 +62,10 @@
 	private int poolShutdownDelay = DEFAULT_POOL_SHUTDOWN_DELAY; 
 	private int pingFrequencySeconds = DEFAULT_PING_FREQ_SECONDS;
 	private int flushTimingsFrequencySeconds = DEFAULT_FLUSH_TIMINGS_FREQ_SECONDS;
-	private boolean localDcAffinity = DEFAULT_LOCAL_DC_AFFINITY;
+	private boolean localZoneAffinity = DEFAULT_LOCAL_DC_AFFINITY;
```
Comment (shailesh33) [inline dyno-core/src/main/java/com/netflix/dyno/connectionpool/impl/ConnectionPoolConfigurationImpl.java]: DEFAULT_LOCAL_ZONE_AFFINITY?

File: dyno-core/src/main/java/com/netflix/dyno/connectionpool/impl/ConnectionPoolImpl.java [hunk 2]
```diff
@@ -284,9 +284,13 @@ public boolean apply(HostConnectionPool<CL> hostPool) {
 		do  {
 			Connection<CL> connection = null;
 			
-			try { 
-					connection = 
-							selectionStrategy.getConnection(op, cpConfiguration.getMaxTimeoutWhenExhausted(), TimeUnit.MILLISECONDS);
+			try {
+					connection = selectionStrategy.getConnectionUsingRetryPolicy(
+                            op,
```
Comment (shailesh33) [inline dyno-core/src/main/java/com/netflix/dyno/connectionpool/impl/ConnectionPoolImpl.java]: Is this a valid indentation

File: dyno-core/src/test/java/com/netflix/dyno/connectionpool/impl/ConnectionPoolImplTest.java [hunk 3]
```diff
@@ -153,7 +153,7 @@ public void beforeTest() {
 			}
 		});
 		
-		cpConfig.setLocalDC("localDC");
+		cpConfig.setLocalRack("localDC");
```
Comment (ipapapa) [inline dyno-core/src/test/java/com/netflix/dyno/connectionpool/impl/ConnectionPoolImplTest.java]: "localRack"?

File: dyno-jedis/src/main/java/com/netflix/dyno/jedis/DynoJedisClient.java [hunk 4]
```diff
@@ -3278,9 +3278,22 @@ public DynoJedisClient build() {
 
             cpConfig.withHostSupplier(hostSupplier);
 
-            if (cpConfig.getTokenSupplier() == null) {
-                Logger.info("TOKEN AWARE selected and no token supplier found, using default HttpEndpointBasedTokenMapSupplier()");
-                cpConfig.withTokenSupplier(new HttpEndpointBasedTokenMapSupplier(port));
+            if (ConnectionPoolConfiguration.LoadBalancingStrategy.TokenAware == cpConfig.getLoadBalancingStrategy()) {
+                if (cpConfig.getTokenSupplier() == null) {
+                    Logger.info("TOKEN AWARE selected and no token supplier found, using default HttpEndpointBasedTokenMapSupplier()");
```
Comment (ipapapa) [inline dyno-jedis/src/main/java/com/netflix/dyno/jedis/DynoJedisClient.java]: Shouldn't this be a warning?

**Comment by ipapapa** [comment]
LGTM

File: dyno-core/src/main/java/com/netflix/dyno/connectionpool/impl/ConnectionPoolConfigurationImpl.java (hunk 1 above)
Comment (jcacciatore) [inline dyno-core/src/main/java/com/netflix/dyno/connectionpool/impl/ConnectionPoolConfigurationImpl.java]: Thanks, renamed to DEFAULT_LOCAL_RACK_AFFINITY

File: dyno-core/src/main/java/com/netflix/dyno/connectionpool/impl/ConnectionPoolImpl.java (hunk 2 above)
Comment (jcacciatore) [inline dyno-core/src/main/java/com/netflix/dyno/connectionpool/impl/ConnectionPoolImpl.java]: I think the tabs messed up the rendering; changed the file indent with spaces now.

File: dyno-core/src/test/java/com/netflix/dyno/connectionpool/impl/ConnectionPoolImplTest.java (hunk 3 above)
Comment (jcacciatore) [inline dyno-core/src/test/java/com/netflix/dyno/connectionpool/impl/ConnectionPoolImplTest.java]: Thanks - changed all localDC string values to localRack

File: dyno-jedis/src/main/java/com/netflix/dyno/jedis/DynoJedisClient.java (hunk 4 above)
Comment (jcacciatore) [inline dyno-jedis/src/main/java/com/netflix/dyno/jedis/DynoJedisClient.java]: Sure - done.

File: dyno-core/src/main/java/com/netflix/dyno/connectionpool/impl/lb/HostSelectionWithFallback.java [hunk 5]
```diff
@@ -163,12 +182,13 @@ public HostSelectionWithFallback(ConnectionPoolConfiguration config, ConnectionP
...
-        return cpConfig.getMaxFailoverCount() > 0 && remoteDCNames.getEntireList().size() > 0;
+        return cpConfig.getMaxFailoverCount() > 0 &&
+                (cpConfig.localZoneAffinity() && remoteDCNames.getEntireList().size() > 0) ||
+                (!cpConfig.localZoneAffinity() && !localSelector.isEmpty());
     }
 
     private HostConnectionPool<CL> getFallbackHostPool(BaseOperation<CL, ?> op, Long token) {
-		
-		int numRemotes = remoteDCNames.getEntireList().size();
+        int numRemotes = remoteDCNames.getEntireList().size();
```
Comment (shailesh33) [inline dyno-core/src/main/java/com/netflix/dyno/connectionpool/impl/lb/HostSelectionWithFallback.java]: is this supposed to be remoteDCNames?

**Pull request merged**